jax/_src/api.py

```diff
@@ -3245,6 +3246,50 @@ def try_to_block(x):
       return x
   return jax.tree_util.tree_map(try_to_block, x)
 
+def pure_callback(callback: Callable[..., Any], result_shape_dtypes: Any,
+                  *args: Any, **kwargs: Any):
+  """Calls a pure Python callback function from staged out JAX programs.
+
+  ``pure_callback`` enables calling a Python function in JIT-ed JAX functions.
+  The input ``callback`` will be passed NumPy arrays in place of JAX arrays and
+  should also return NumPy arrays. Execution takes place on the CPU host.
+
+  The callback is treated as "pure" meaning it can be called multiple times when
+  transformed (for example in a ``vmap`` or ``pmap``), and it can also
+  potentially be removed from JAX programs via dead-code elimination. Pure
+  callbacks can also be reordered if data-dependence allows.
+
+  When ``pmap``-ed, the pure callback will be called several times (one on each axis
+  of the map). When `vmap`-ed the behavior will depend on the value of the
+  ``rank_polymorphic`` keyword argument. If the callback is indicated as rank
+  polymorphic, the callback will be called directly on batched inputs (where the
+  batch axis is the leading dimension). Additionally, the callbacks should
+  return outputs that also have a leading batch axis. If not rank polymorphic,
+``callback`` will be mapped sequentially across the batched axis.
+
+  Args:
+    callback: A Python callable. The callable will be passed in NumPy arrays and
+      should return a PyTree of NumPy arrays that matches
+      ``result_shape_dtypes``.
+    result_shape_dtypes: A PyTree of Python objects that have ``shape`` and
+      ``dtype`` properties that correspond to the shape and dtypes of the
+      outputs of ``callback``.
+    *args: The positional arguments to the callback. Must be PyTrees of JAX
+      types.
+    rank_polymorphic: A boolean that indicates whether or not ``callback`` is
```

Comment:
See above suggestion about renaming `rank_polymorphic` (and avoiding that terminology).